promql/engine.go

```diff
@@ -1325,16 +1395,100 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
+		subqueryEvalTime := timestamp.FromFloatSeconds(e.Timestamp)
+		subqueryStartTime := ev.startTimestamp - offsetMillis - rangeMillis
+		if subqueryEvalTime != 0 {
+			newEv.endTimestamp = subqueryEvalTime - offsetMillis
+			subqueryStartTime = newEv.endTimestamp - rangeMillis
+		}
+
+		// Start with the first timestamp after subqueryStartTime
 		// that is aligned with the step (multiple of 'newEv.interval').
-		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
-		if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
+		newEv.startTimestamp = newEv.interval * (subqueryStartTime / newEv.interval)
+		if newEv.startTimestamp < subqueryStartTime {
 			newEv.startTimestamp += newEv.interval
 		}
 
 		res, ws := newEv.eval(e.Expr)
 		ev.currentSamples = newEv.currentSamples
 		return res, ws
+	case *parser.StepInvariantExpr:
+		switch ce := e.Expr.(type) {
+		case *parser.StringLiteral, *parser.NumberLiteral:
+			return ev.eval(ce)
+		}
+
+		if e.Result != nil {
+			return e.Result, nil
+		}
+		newEv := &evaluator{
+			startTimestamp:           ev.startTimestamp,
+			endTimestamp:             ev.startTimestamp,
+			interval:                 ev.interval,
+			ctx:                      ev.ctx,
+			currentSamples:           ev.currentSamples,
+			maxSamples:               ev.maxSamples,
+			logger:                   ev.logger,
+			lookbackDelta:            ev.lookbackDelta,
+			noStepSubqueryIntervalFn: ev.noStepSubqueryIntervalFn,
+		}
+
+		canExtrapolate := canExtrapolateStepInvariantExpr(e.Expr)
+		if !canExtrapolate {
+			newEv.endTimestamp = ev.endTimestamp
+		}
+
+		res, ws := newEv.eval(e.Expr)
+		ev.currentSamples = newEv.currentSamples
+
+		_, isSubquery := e.Expr.(*parser.SubqueryExpr)
+		if canExtrapolate && !isSubquery {
+			switch val := res.(type) {
+			case Matrix:
+				for i := range val {
+					if len(val[i].Points) != 1 {
+						panic(errors.Errorf("unexpected number of samples"))
+					}
+					for ts := ev.startTimestamp + ev.interval; ts <= ev.endTimestamp; ts = ts + ev.interval {
+						val[i].Points = append(val[i].Points, Point{
+							T: ts,
+							V: val[i].Points[0].V,
+						})
+						ev.currentSamples++
+						if ev.currentSamples >= ev.maxSamples {
+							ev.error(ErrTooManySamples(env))
+						}
+					}
+				}
+			case Vector:
+				mat := make(Matrix, len(val))
+				for i, smpl := range val {
+					mat[i] = Series{
+						Metric: smpl.Metric,
+						Points: make([]Point, 0, 1+(ev.endTimestamp-ev.startTimestamp)/ev.interval),
+					}
+					for ts := ev.startTimestamp + ev.interval; ts <= ev.endTimestamp; ts = ts + ev.interval {
+						mat[i].Points = append(mat[i].Points, Point{
+							T: ts,
+							V: smpl.V,
+						})
+						ev.currentSamples++
+						if ev.currentSamples >= ev.maxSamples {
+							ev.error(ErrTooManySamples(env))
+						}
+					}
+				}
+				res = mat
+			case Scalar:
+				// Do nothing.
```

Comment:
This doesn't seem right at all, these should be handled like the others.

Reply:
How can we return an array of `Scalar` type?

Reply:
Also I don't see the timestamp of the scalar being used anywhere except for printing, it's only the value.

Reply:
The timestamp isn't relevant, the question is why is this doing nothing? Scalar expressions are a thing.

Reply:
Do nothing because return the result as it is.

Reply:
The comment could make that clearer.

Reply:
Ack